Use SSI for Style blocks in web pages; suppress log messages routine to testing; suppress routine messages during ant build

bobjacobsen opened the pull request:
Goal is to reduce the volume of the Travis build output.  

The JDNS logging completely dominates the middle, but this simplifies/shortens the build text at the beginning and spurious log messages at the end.

pabender commented:
Is there a library we need to add to the build system for change to work?

After the addition of the quite macro, I get an error because of the javascript contained in the macro when running the `ant debug` target.  Here's the most recent error:

/home/Paul/JMRI/git/JMRI/build.xml:277: The following error occurred while executing this line:
/home/Paul/JMRI/git/JMRI/build.xml:220: java.lang.NoClassDefFoundError: org/mozilla/javascript/EvaluatorException
	at java.lang.Class.getDeclaredConstructors0(Native Method)
	at java.lang.Class.privateGetDeclaredConstructors(Class.java:2671)
	at java.lang.Class.getConstructor0(Class.java:3075)
	at java.lang.Class.newInstance(Class.java:412)
	at org.apache.bsf.BSFManager.loadScriptingEngine(BSFManager.java:697)
	at org.apache.bsf.BSFManager.exec(BSFManager.java:485)
	at org.apache.tools.ant.util.optional.ScriptRunner.executeScript(ScriptRunner.java:100)
	at org.apache.tools.ant.taskdefs.optional.Script.execute(Script.java:52)
	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
	at org.apache.tools.ant.Task.perform(Task.java:348)
	at org.apache.tools.ant.taskdefs.Sequential.execute(Sequential.java:68)
	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
	at org.apache.tools.ant.Task.perform(Task.java:348)
	at org.apache.tools.ant.taskdefs.MacroInstance.execute(MacroInstance.java:396)
	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
	at org.apache.tools.ant.Task.perform(Task.java:348)
	at org.apache.tools.ant.Target.execute(Target.java:435)
	at org.apache.tools.ant.Target.performTasks(Target.java:456)
	at org.apache.tools.ant.Project.executeSortedTargets(Project.java:1393)
	at org.apache.tools.ant.Project.executeTarget(Project.java:1364)
	at org.apache.tools.ant.helper.DefaultExecutor.executeTargets(DefaultExecutor.java:41)
	at org.apache.tools.ant.Project.executeTargets(Project.java:1248)
	at org.apache.tools.ant.Main.runBuild(Main.java:851)
	at org.apache.tools.ant.Main.startAnt(Main.java:235)
	at org.apache.tools.ant.launch.Launcher.run(Launcher.java:280)
	at org.apache.tools.ant.launch.Launcher.main(Launcher.java:109)
Caused by: java.lang.ClassNotFoundException: org.mozilla.javascript.EvaluatorException
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 38 more

rhwood commented:
There shouldn't be. The Nashorn JavaScript engine is built into Java 1.8, and the Mozilla one into earlier versions. You may need to wrap the script in CDATA blocks (like at https://github.com/JMRI/JMRI/blob/docbook/doc/build.xml#L80).

rhwood commented:
> Why didn't we do this earlier? IIRC, there was a reason to want a more static page layout, but I no
> longer recall the reason… Was it some sort of doc tree/online help as part of the install?

We don't want to do this because:
- It's hostile to the JavaHelp, which is not an SSI-capable engine
- It can't be shared using the built in web server in JMRI, which is not SSI-capable
- It's not portable; GitHub's file sharing, for instance, is not SSI-capable
- Others providing quick edits need to setup a complete SSI-capable server to see their changes in context, and can't just open the file being edited after having run an ant task to assemble it

bobjacobsen commented:
> On Sep 26, 2015, at 10:19 PM, Paul Bender <notifications@github.com> wrote:
> 
> Is there a library we need to add to the build system for change to work?
> 
> After the addition of the quite macro, I get an error because of the javascript contained in the macro when running the ant debug target. Here's the most recent error:
> /home/Paul/JMRI/git/JMRI/build.xml:277: The following error occurred while executing this line:
> /home/Paul/JMRI/git/JMRI/build.xml:220: java.lang.NoClassDefFoundError: org/mozilla/javascript/EvaluatorException
> at java.lang.Class.getDeclaredConstructors0(Native Method)

(This is about the “quiet” element in the Ant build.xml file; there’s a separate part of the thread about SSI and Style)

Perhaps there’s something about the Ant version?

I’m using 

 % ant -version
Apache Ant(TM) version 1.9.2 compiled on July 8 2013

on Mac OS X. I think it was installed by XCode (I have several 7+ versions installed, not sure which was last to update the command-line tools)

Randall wrote:
> There shouldn't be. The Nashorn JavaScript engine is built into Java 1.8, and the Mozilla one into earlier versions. You may need to wrap the script in CDATA blocks (like at https://github.com/JMRI/JMRI/blob/docbook/doc/build.xml#L80).

That might be a clue.  Paul’s Ant is complaining of the lack of the Mozilla one.  Java 1.8 should be looking for Nashorn. Perhaps there’s a mixed version problem, with an older .jar file still around in ~/.ant/lib or similar?

Bob

--
Bob Jacobsen
jacobsen@mac.com +1-510-486-7355 AIM, Skype JacobsenRG

bobjacobsen commented:
(Referring to the use of SSI to access the Style block from a common file)

> On Sep 27, 2015, at 3:44 AM, Randall Wood <notifications@github.com> wrote:
> 
> Why didn't we do this earlier? IIRC, there was a reason to want a more static page layout, but I no
> longer recall the reason… Was it some sort of doc tree/online help as part of the install?
> 
> We don't want to do this because:
> 
> 	• It's hostile to the JavaHelp, which is not an SSI-capable engine
> 	• It can't be shared using the built in web server in JMRI, which is not SSI-capable
> 	• It's not portable; GitHub's file sharing, for instance, is not SSI-capable
> 	• Others providing quick edits need to setup a complete SSI-capable server to see their changes in context, and can't just open the file being edited after having run an ant task to assemble it

We _already_ use SSI for the Header, Footer and where needed, the Sidebar, parts of the web pages.  

http://jmri.sourceforge.net/help/en/html/doc/Technical/WebSite.shtml

That has two important advantages:

1) It makes it practical to maintain those common parts of the web site.
2) It allows us to use common files for JavaHelp and other uses. Since JavaHelp doesn’t interpret SSI, the use of SSI keeps Header/Footer/Sidebar web-site chrome away from the JavaHelp display.

The only downside I’ve encountered is that sometimes people submit patches based on the expanded pages they’ve downloaded.  With GitHub direct editing, perhaps we’ll see that less. In any case, it’s easy to fix because of the <!-- Header --><!-- /Header —> comments.

A long time ago, JohnP set up the web site so that the Header, Footer, etc files where merged via an update.pl script that was to be run over the enter site when a change was made.  That file is still there, but as soon as SSI was available on the SF.net servers we switched to using that.

The Style file contains:

  <META http-equiv=Content-Type content="text/html; charset=iso-8859-1">
  <link rel="stylesheet" type="text/css" href="/css/default.css" media="screen">
  <link rel="stylesheet" type="text/css" href="/css/print.css" media="print">
  <link rel="icon" href="/images/jmri.ico" type="image/png">
  <link rel="home" title="Home" href="/">

AFAIK, none of those are relevant for JavaHelp. 

http://jmri.sourceforge.net/help/en/html/doc/Technical/Help.shtml

JavaHelp doesn’t use CSS (It’s described as  “Plain old HTML 4.0”, but there’s otherwise not a lot of detail available)

http://jmri.sourceforge.net/help/en/html/doc/Technical/WebSite.shtml


It’s possible that some of those other uses of the help/ tree do need those lines, but don’t do SSI.  I don’t know of any, but they could exist.  So I converted some pages that are accessed from the help (the tables pages) and from the web (the doc/Technical pages) to see if I could flush out issues.

Bob

--
Bob Jacobsen
jacobsen@mac.com +1-510-486-7355 AIM, Skype JacobsenRG

pabender commented:
On 09/27/2015 10:41 AM, Bob Jacobsen wrote:
> > On Sep 26, 2015, at 10:19 PM, Paul Bender <notifications@github.com> 
> wrote:
> >
> > Is there a library we need to add to the build system for change to 
> work?
> >
> > After the addition of the quite macro, I get an error because of the 
> javascript contained in the macro when running the ant debug target. 
> Here's the most recent error:
> > /home/Paul/JMRI/git/JMRI/build.xml:277: The following error occurred 
> while executing this line:
> > /home/Paul/JMRI/git/JMRI/build.xml:220: 
> java.lang.NoClassDefFoundError: org/mozilla/javascript/EvaluatorException
> > at java.lang.Class.getDeclaredConstructors0(Native Method)
>
> (This is about the “quiet” element in the Ant build.xml file; there’s 
> a separate part of the thread about SSI and Style)
>
> Perhaps there’s something about the Ant version?
>
> I’m using
>
> % ant -version
> Apache Ant(TM) version 1.9.2 compiled on July 8 2013
>

I'm using:
Apache Ant(TM) version 1.9.2 compiled on September 12 2013

This is on a Fedora machine.

> Randall wrote:
> > There shouldn't be. The Nashorn JavaScript engine is built into Java 
> 1.8, and the Mozilla one into earlier versions. You may need to wrap 
> the script in CDATA blocks (like at 
> https://github.com/JMRI/JMRI/blob/docbook/doc/build.xml#L80).
>
> That might be a clue. Paul’s Ant is complaining of the lack of the 
> Mozilla one. Java 1.8 should be looking for Nashorn. Perhaps there’s a 
> mixed version problem, with an older .jar file still around in 
> ~/.ant/lib or similar?

I'm using java 1.8.0_45.  That's probably not the most up to date java, 
but it shouldn't be an issue.

I don't even have a ~/.ant for this user on this machine....

Hopefully your changes in pull request 43 will fix this.  Once those get 
merged in I'll pull the repository and try again.

Paul

bobjacobsen commented:
(re-open as Paul still having trouble)

rhwood commented:
This is being tracked in #43
It appears that including an Eclipse plugin via RPM caused this error.
